include/lsst/meas/astrom/SipTransform.h

```diff
@@ -0,0 +1,357 @@
+        _cdMatrix(cdMatrix),
+        _poly(poly)
+    {}
+
+    SipTransformBase(SipTransformBase const & other) = default;
+    SipTransformBase(SipTransformBase && other) = default;
+    SipTransformBase & operator=(SipTransformBase const & other) = default;
+    SipTransformBase & operator=(SipTransformBase && other) = default;
+
+    void swap(SipTransformBase & other) {
+        std::swap(_pixelOrigin, other._pixelOrigin);
+        std::swap(_cdMatrix, other._cdMatrix);
+        _poly.swap(other._poly);
+    }
+
+    afw::geom::Point2D _pixelOrigin;
+    afw::geom::LinearTransform _cdMatrix;
+    PolynomialTransform _poly;
+};
+
+
+/**
+ *  A transform that maps pixel coordinates to intermediate world coordinates
+ *  according to the SIP convention.
+ *
+ *  The SIP forward transform is defined as
+ *  @f[
+ *     \left[\begin{array}{ c }
+ *       x \\
+ *       y
+ *     \end{array}\right]
+ *     = \mathbf{Z}
+ *     \left[\begin{array}{ c }
+ *       (u - u_0) + {\displaystyle\sum_{p,q}^{2 \le p + q \le N}} \mathrm{A}_{p,q} (u-u_0)^p (v-v_0)^q \\
+ *       (v - v_0) + {\displaystyle\sum_{p,q}^{2 \le p + q \le N}} \mathrm{B}_{p,q} (u-u_0)^p (v-v_0)^q
+ *     \end{array}\right]
+ *  @f]
+ *  where
+ *   - @f$(u,v)@f$ are pixel coordinates.
+ *   - @f$(x,y)@f$ are "intermediate world coordinates" -- the result of
+ *     applying the gnomonic (TAN) projection at sky origin CRVAL to sky
+ *     coordinates).
+ *   - @f$\mathbf{Z}@f$ is the @f$2 \times 2@f$ linear transform (@f$\mathrm{CD}@f$) matrix.
+ *   - @f$(u_0,v_0)@f$ is the pixel origin @f$\mathrm{CRVAL}@f$.
+ *   - @f$\mathrm{A}@f$, @f$\mathrm{B}@f$ are the polynomial coefficients of
+ *     the forward transform.
+ *
+ *  The SIP convention encourages (but does not require) nulling the zeroth-
+ *  and first-order elements of
+ *  @f$\mathrm{A}@f$ and @f$\mathrm{B}@f$, which ensures the representation of
+ *  a given transform is unique.  This also makes fitting a SIP transform to
+ *  data a nonlinear operation, as well as making the conversion from standard
+ *  polynomial transforms to SIP form impossible in general.
+ */
+class SipForwardTransform : public SipTransformBase {
+public:
+
+    /**
+     *  Convert a PolynomialTransform to an equivalent SipForwardTransform.
+     *
+     *  @param[in]   poly           PolynomialTransform to convert.
+     *  @param[in]   pixelOrigin    CRPIX @f$(u_0,v_0)@f$
+     *  @param[in]   cdMatrix       CD matrix @f$Z@f$
+     */
+    static SipForwardTransform convert(
+        PolynomialTransform const & poly,
+        afw::geom::Point2D const & pixelOrigin,
+        afw::geom::LinearTransform const & cdMatrix
+    );
+
+    /**
+     *  Convert a ScaledPolynomialTransform to an equivalent SipForwardTransform.
+     *
+     *  @param[in]   scaled         ScaledPolynomialTransform to convert.
+     *  @param[in]   pixelOrigin    CRPIX @f$(u_0,v_0)@f$
+     *  @param[in]   cdMatrix       CD matrix @f$Z@f$
```

Comment:
I did at one point, but that was before a lot of copy-and-pasting, and this is a good reminder to do a final check before I merge.